bech32/bech32.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2017 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package bech32
+
+import (
+	"fmt"
+	"strings"
+)
+
+const charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+var gen = []int{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}
+
+// Decode decodes a bech32 encoded string, returning the human-readable
+// part and the data part excluding the checksum.
+func Decode(bech string) (string, []byte, error) {
+	// Only	ASCII characters between 33 and 126 are allowed.
+	for i := 0; i < len(bech); i++ {
+		if bech[i] < 31 || bech[i] > 126 {
+			return "", nil, fmt.Errorf("invalid character in " +
+				"string")
+		}
+	}
+
+	// The characters must be either all lowercase or all uppercase.
+	lower := strings.ToLower(bech)
+	upper := strings.ToUpper(bech)
+	if bech != lower && bech != upper {
+		return "", nil, fmt.Errorf("string not all lowercase or all " +
+			"uppercase")
+	}
+
+	// We'll work with the lowercase string from now on.
+	bech = lower
+
+	// The string is invalid if the last '1' is non-existent, it is the
+	// first character of the string (no human-readable part) or one of the
+	// last 6 characters of the string (since checksum cannot contain '1'),
+	// or if the string is more than 90 characters in total.
+	one := strings.LastIndexByte(bech, '1')
+	if one < 1 || one+7 > len(bech) {
+		return "", nil, fmt.Errorf("invalid index of 1")
+	}
+	if len(bech) > 90 {
+		return "", nil, fmt.Errorf("string too long")
+	}
+
+	// The human-readable part is everything before the last '1'.
+	hrp := bech[:one]
+	data := bech[one+1:]
+
+	// Each character corresponds to the byte with value of the index in
+	// 'charset'.
+	decoded := make([]byte, 0, len(data))
+	for i := 0; i < len(data); i++ {
+		index := strings.IndexByte(charset, data[i])
+		if index < 0 {
+			return "", nil, fmt.Errorf("invalid character in " +
+				"data part")
+		}
+		decoded = append(decoded, byte(index))
+	}
+
+	if !bech32VerifyChecksum(hrp, decoded) {
+		return "", nil, fmt.Errorf("checksum failed")
+	}
+
+	// We exclude the last 6 bytes, which is the checksum.
+	return hrp, decoded[:len(decoded)-6], nil
+}
+
+// Encode encodes a byte slice into a bech32 string with the
+// human-readable part hrb. Note that the bytes must each encode 5 bits
+// (base32).
+func Encode(hrp string, data []byte) (string, error) {
+	// Calculate the checksum of the data and append it at the end.
+	checksum := bech32Checksum(hrp, data)
+	combined := append(data, checksum...)
```

Comment:
I would prefer this not be done here.  It's an extra unnecessary allocation.  Just append the data and checksum directly to the result, which is already properly sized below and replace the encoded characters directly in place.

e.g.

```Go
dataOffset := len(hrp)+1
checksum := bech32Checksum(hrp, data)
result := make([]byte, 0, len(hrp)+1+len(data)+len(checksum))
result = append(result, hrp...)
result = append(result, '1')
result = append(result, data...)
result = append(result, checksum...)
for i, b := range result[dataOffset:] {
		if int(b) >= len(charset) {
			fmt.Printf("invalid data byte: %v", b)
			return
		}
		result[dataOffset+i] = charset[b]
}
```